core/unix/signal.c

```diff
@@ -7710,6 +7708,16 @@ os_forge_exception(app_pc target_pc, dr_exception_type_t type)
      * to a plain frame on delivery.
      */
     memset(frame, 0, sizeof(*frame));
+
+    kernel_ucontext_t *uc = get_ucontext_from_rt_frame(frame);
+#if defined(MACOS)
+    /* Since SIGCXT_FROM_UCXT just accesses the uc->uc_mcontext ptr field on
+     * macOS, sc will be NULL below if we do not initialize uc_mcontext first
+     */
+    uc->IF_X64_ELSE(uc_mcontext64, uc_mcontext) = (void *)&frame->mc;
```

Comment:
This seems weird.
`frame` is zeroed-out anyway, isn't this just changing `sc` to be a "different" NULL?
Also, shouldn't `sc` being NULL break more than MACOS?

Reply:
> frame is zeroed-out anyway,

frame is indeed zeroed out; which means frame->uc.uc_mcontext is NULL before L7717.

> isn't this just changing sc to be a "different" NULL?

L7717 seems to set  frame->uc.uc_mcontext (which was NULL before) to &frame->mc; the address of mc within the frame is not NULL.

I'm not familiar with these Mac structs. I wonder if it is simply a requirement for frame->uc.uc_mcontext to point to frame->mc.

Reply:
@ndrewh The comment at L7714 justifies that `uc->IF_X64_ELSE(uc_mcontext64, uc_mcontext)` must be set to something, but it doesn't explain why `&frame->mc` is the correct intended value. Can you add that reasoning to the comment?

Reply:
> I'm not familiar with these Mac structs. I wonder if it is simply a requirement for frame->uc.uc_mcontext to point to frame->mc.

i am not an expert and the XNU code is a bit convoluted here but i think the answer is yes...

https://github.com/apple-oss-distributions/xnu/blob/8d741a5de7ff4191bf97d57b9f54c2f6d4a15585/bsd/dev/arm/unix_signal.c#L411

fixing comment